File: src/main/java/core/scm/MapEntry.java
package core.scm;

import core.procedures.FnArgsBuilder;
import core.writer.Writer;

import java.util.Map;

public class MapEntry extends Vector implements Map.Entry {

  private final Object key;
  private final Object value;

  public MapEntry(Object key, Object value) {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Type.ExactNonNegativeInteger.class}).build());
    this.key = key;
    this.value = value;
  }

  @Override
  public Object getKey() {
    return key;
  }

  @Override
  public Object getValue() {
    return value;
  }

  @Override
  public Object setValue(Object value) {
    throw new UnsupportedOperationException();
  }

  @Override
  public String getName() {
    return "map entry";
  }

  @Override
  public String toString() {
    return "[" + Writer.write(key) + " " + Writer.write(value) + "]";
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    MapEntry that = (MapEntry) o;
    if (key != null ? !key.equals(that.key) : that.key != null) return false;
    return value != null ? value.equals(that.value) : that.value == null;
  }

  @Override
  public int hashCode() {
    int result = key != null ? key.hashCode() : 0;
    result = 31 * result + (value != null ? value.hashCode() : 0);
    return result;
  }
}

File: src/main/java/core/scm/MapEntry.kt
package core.scm

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.writer.Writer

class MapEntry(override var key: Any?, override var value: Any?) :
        AFn<Number?, Any?>(minArgs = 1, maxArgs = 1, isPure = true, mandatoryArgsTypes = arrayOf(Int::class.java)),
        Map.Entry<Any?, Any?>, IAssoc<Number?, Any?>, Sequence<Any?> {

    constructor(entry: Map.Entry<Any?, Any?>) : this(entry.key, entry.value)

    override fun containsKey(key: Number?) = when {
        key == null -> false
        key.toInt() == 0 -> true
        key.toInt() == 1 -> true
        else -> false
    }

    override fun getEntry(key: Number?) = when {
        key == null -> null
        key.toInt() == 0 -> MapEntry(0, this.key)
        key.toInt() == 1 -> MapEntry(1, this.value)
        else -> null
    }

    override fun assoc(key: Number?, value: Any?): MapEntry {
        Type.assertType("Map Entry", key, Int::class.java)
        when {
            key == null -> throw WrongTypeException("Map Entry", "Integer", key)
            key.toInt() == 0 -> this.key = value
            key.toInt() == 1 -> this.value = value
            else -> throw IndexOutOfBoundsException()
        }
        return this
    }

    override operator fun invoke(arg: Number?) = getEntry(arg)

    override fun iterator() = sequenceOf(key, value).iterator()

    override fun toString() = "[${Writer.write(key)} ${Writer.write(value)}]"

    override fun equals(other: Any?) = when {
        this === other -> true
        other == null || javaClass != other.javaClass -> false
        else -> key == (other as MapEntry).key && value == other.value
    }

    override fun hashCode() = 31 * (key?.hashCode() ?: 0) + (value?.hashCode() ?: 0)
}
